src/linker/Linker.Steps/RootAssemblyInputStep.cs

```diff
@@ -55,15 +55,27 @@ protected override void Process ()
 				Annotations.AddPreservedMethod (ep.DeclaringType, ep);
 				break;
 			case AssemblyRootMode.VisibleMembers:
-				var preserve = HasInternalsVisibleTo (assembly) ? TypePreserveMembers.AllVisibleOrInternal : TypePreserveMembers.AllVisible;
+				var preserve_visible = TypePreserveMembers.Visible;
+				if (HasInternalsVisibleTo (assembly))
+					preserve_visible |= TypePreserveMembers.Internal;
 
-				var module = assembly.MainModule;
-				if (module.HasExportedTypes)
-					foreach (var type in module.ExportedTypes)
-						MarkAndPreserveVisible (assembly, type, preserve);
+				MarkAndPreserve (assembly, preserve_visible);
+				break;
+
+			case AssemblyRootMode.Library:
+				var preserve_library = TypePreserveMembers.Visible | TypePreserveMembers.Library;
+				if (HasInternalsVisibleTo (assembly))
+					preserve_library |= TypePreserveMembers.Internal;
+
+				MarkAndPreserve (assembly, preserve_library);
 
-				foreach (var type in module.Types)
-					MarkAndPreserveVisible (type, preserve);
+				// Assembly root mode wins over any enabled optimization which
+				// could conflict with library rooting behaviour
```

Comment:
Future refactoring idea: If assembly action is not Link we also effectively disable lot of optimizations, but it's done in each usage site explicitly (check for action and then check for the optimization). Maybe it would be cleaner to disable optimization based on actions as well (not just root mode).

Reply:
Agreed - we are increasingly blurring the line between actions and root modes and it seems worthwhile to consider unifying them or separating out the overlapping behaviors.